llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -438,6 +438,73 @@ bool duplicateAdressingRegs(MachineBasicBlock &MBB, MachineRegisterInfo &MRI) {
   return tryToDuplicateLoadUse(LoadUses, NonLoadUses, MRI, TII);
 }
 
+/// This optimization tries to propagate GlobalValues as MMO of load operations
+/// that don't use normal pointers as a source of an address.
+void findGlobalValues(const MachineInstr &MI,
+                      SmallPtrSet<const Value *, 4> &GVSet,
+                      SmallPtrSet<const MachineInstr *, 4> &VisitedInstrs,
+                      unsigned RecMax, MachineRegisterInfo &MRI) {
+
+  // Exhausted the search.
+  if (RecMax == 0)
+    return;
+
+  // Loaded values cannot be tracked as Globals.
+  if (MI.mayLoad())
+    return;
+
+  if (VisitedInstrs.find(&MI) != VisitedInstrs.end())
+    return;
+
+  VisitedInstrs.insert(&MI);
+
+  for (auto MO : MI.uses()) {
+    if (MO.isGlobal()) {
+      GVSet.insert(MO.getGlobal());
+    } else if (MO.isReg() && MO.getReg().isVirtual()) {
```

Comment:
I think you should recognize the regular address modifiers here and stop recursion. You don't want to descend into integer operands of e.g. GEP.

Reply:
Hi @martien-de-jong, as we are in MIR now, all GEPs were translated to ptradd, mul and etc.

Reply:
GEP, PTRADD, not much different. PTR_ADD only leads to pointers in its first operand, not in the second, so searching the second is wasted.

Reply:
But what if the first operand is null  (or something like G_INTTOPTR(G_CONSTANT))  and second is an integer derived from a  G_PTRTOINT (or something like G_PTRTOINT(GlobalValue))? Pointer arithmetic can be crazy... like:

```
body:             |
  bb.1.entry:
    %1:_(p0) = G_CONSTANT i20 4
    %4:_(p0) = G_GLOBAL_VALUE @Global
    %3:_(s20) = G_PTRTOINT %4(p0)
    %7:_(s20) = G_CONSTANT i20 2
    %5:_(s20) = G_SHL %3, %7(s20)
    %6:_(p0) = G_PTR_ADD %1, %5(s20)
    $p0 = COPY %6(p0)
    PseudoRET implicit $lr, implicit $p0

...
```

In this example we can reach the pointer though the second operand.

Reply:
I wouldn't mind if those programmers were punished...
But seriously, if we descend through any operator, the recursion depth is not nicely modelling processing effort. I would rather have somethin linear like number of visited nodes.

Reply:
As discussed offline, the recursion limit was replaced by the number of visited instructions.